test/integration/podgc/podgc_test.go

```diff
@@ -0,0 +1,200 @@
+	}
+
+	err = wait.PollImmediate(time.Second, time.Second*15, func() (bool, error) {
+		updatedPod, err := cs.CoreV1().Pods(testCtx.NS.Name).Get(testCtx.Ctx, pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return true, err
+		}
+		if updatedPod.ObjectMeta.DeletionTimestamp != nil {
+			return true, nil
+		}
+		return false, nil
+	})
+	if err != nil {
+		t.Fatalf("Error '%v' while waiting for the pod '%v' to be deleted", err, klog.KObj(pod))
+	}
+}
+
+// TestTerminatingOnOutOfServiceNode tests deletion pods terminating on out-of-service nodes
+func TestTerminatingOnOutOfServiceNode(t *testing.T) {
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.NodeOutOfServiceVolumeDetach, true)()
+	testCtx := setup(t, "podgc-out-of-service")
+	defer testutils.CleanupTest(t, testCtx)
+	cs := testCtx.ClientSet
+
+	node := &v1.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "node",
+			Namespace: testCtx.NS.Name,
+		},
+		Spec: v1.NodeSpec{},
+		Status: v1.NodeStatus{
+			Conditions: []v1.NodeCondition{
+				{
+					Type:   v1.NodeReady,
+					Status: v1.ConditionFalse,
+				},
+			},
+		},
+	}
+	node, err := cs.CoreV1().Nodes().Create(testCtx.Ctx, node, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("Failed to create node '%v', err: %v", node.Name, err)
+	}
+
+	pod := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "testpod",
+			Namespace: testCtx.NS.Name,
+		},
+		Spec: v1.PodSpec{
+			NodeName: node.Name,
+			Containers: []v1.Container{
+				{Name: "foo", Image: "bar"},
+			},
+		},
+	}
+
+	pod, err = cs.CoreV1().Pods(testCtx.NS.Name).Create(testCtx.Ctx, pod, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("Error %v, while creating pod: %v", err, klog.KObj(pod))
+	}
+
+	// trigger termination of the pod, but with long grace period so that it is not removed immediately
+	err = cs.CoreV1().Pods(testCtx.NS.Name).Delete(testCtx.Ctx, pod.Name, metav1.DeleteOptions{})
+	if err != nil {
+		t.Fatalf("Error: '%v' while deleting pod: '%v'", err, klog.KObj(pod))
+	}
+
+	// get the pod status to make sure the pod is not deleted yet
+	pod, err = cs.CoreV1().Pods(pod.Namespace).Get(testCtx.Ctx, pod.Name, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Error: '%v' while checking the pod '%v' is not deleted yet", err, klog.KObj(pod))
+	}
+
+	// taint the node with the out-of-service taint
+	err = testutils.AddTaintToNode(cs, pod.Spec.NodeName, v1.Taint{Key: v1.TaintNodeOutOfService, Value: "", Effect: v1.TaintEffectNoExecute})
+	if err != nil {
+		t.Fatalf("Failed to taint node: %v, err: %v", pod.Spec.NodeName, err)
+	}
+
+	err = wait.PollImmediate(time.Second, time.Second*15, func() (bool, error) {
+		updatedPod, err := cs.CoreV1().Pods(pod.Namespace).Get(testCtx.Ctx, pod.Name, metav1.GetOptions{})
+		if err != nil && !apierrors.IsNotFound(err) {
+			return true, err
+		}
+		return updatedPod == nil || (err != nil && apierrors.IsNotFound(err)), nil
```

Comment:
can we break this down in different if statements so is easier to read?
I think that we are looking for something like, but  please correct me if I'm wrong, I'm not familiar with the conditions we are looking for in this feature
```
if err == nil  {
    return  updatedPod == nil, nil
}
// there was an error
if apierrors.IsNotFound(err) {
    return true, nil
}
return true, err

```

Reply:
updated